### DISCUSSION TITLE
연관된 모델을 어떻게 가져올 것인가?

### DISCUSSION THREAD

**Discussion by AlbertImKr** [weaverse-techtide/weaverse-backend#93] [General]
# Parent 모델을 가져오고 Child 모델을 하나씩 가져오는 방법

- Course 조회: 1
- Course의 Lecture 조회: 1
- Lecture의 Topic 조회: 1
- Topic의 Video 조회: 1

# 단일 쿼리로 연관된 모델 조회하는 방법

- Course 모델과 연관된 모든 데이터를 한 번의 쿼리로 가져오는 방법

**Comment by devnproyj22** [comment]
| 구현 방법 | 장점 | 단점 | 최적화 수준 | 구현 복잡도 | 적합한 상황 |
|--------|----------|------|-------------|-------------|-------------|
| 시리얼라이저 (SerializerMethodField) | - 유연한 데이터 구조<br>- 복잡한 로직 구현 가능 | - 성능 이슈 가능성<br>- N+1 쿼리 문제 | 낮음 | 중간 | 복잡한 데이터 변환이 필요한 경우 |
| 모델 매니저 | - 쿼리 최적화 용이<br>- 재사용성 높음 | - 모델 로직이 복잡해질 수 있음 | 높음 | 중간 | 반복적인 복잡한 쿼리가 필요한 경우 |
| 페이지네이션 | - 대량 데이터 처리 가능<br>- 응답 시간 개선 | - 전체 데이터를 한 번에 볼 수 없음 | 중간 | 낮음 | 대규모 데이터셋을 다룰 때 |
| 선택적 데이터 로드 | - 필요한 데이터만 로드<br>- 성능 향상 | - 추가 쿼리 발생 가능<br>- 구현 복잡도 증가 | 중간 | 높음 | 특정 상황에 따라 다른 데이터가 필요할 때 |
| 중첩된 시리얼라이저 | - 직관적인 데이터 구조<br>- DRF 기본 기능 활용 | - 깊은 중첩 시 성능 저하<br>- 과도한 데이터 로드 | 중간 | 낮음 | 관계가 명확하고 일관된 데이터 구조일 때 |

**Comment by Yoo117** [comment]
홍광님이 말씀하셨던 Django ORM의 Prefetch가 괜찮을 것 같네요
예를 들어 video와 연관된 topic, lecture, course 정보를 가져오고 싶다면

```
from django.db.models import Prefetch

videos = Video.objects.select_related('topic__lecture__course').all()
```

이런 식으로도 쓸 수 있다고도 하더라구요.
